### PR TITLE
Use aligned_alloc instead of malloc for aligned memory allocation.

### PR DESCRIPTION
**Pull request by kimshrier**
The code in lib/std/heap.zig uses malloc for a memory allocation that needs to be aligned to the requirements of std.c.max_align_t.  Malloc makes no guarantee about the alignment of memory that it returns.  The existing code does a run-time check to make sure that the proper alignment has been achieved and panics if it is not properly aligned.

The means that memory allocation at this point in the code works by accident.  This is something I would think is undesirable.

There is a standard memory allocation function, aligned_alloc, that does exactly what is needed.  The only caveat is that the length of the allocated memory needs to be a multiple of the alignment.  Behavior is undefined if this is not true.  That is the reason that I adjust the value of len if it isn't already a multiple of the alignment.

### CODE REVIEW

File: lib/std/heap.zig [hunk 1]
```diff
@@ -182,7 +182,10 @@ fn rawCAlloc(
     _ = ret_addr;
     assert(log2_ptr_align <= comptime std.math.log2_int(usize, @alignOf(std.c.max_align_t)));
     // TODO: change the language to make @ptrCast also do alignment cast
-    const ptr = @alignCast(@alignOf(std.c.max_align_t), c.malloc(len));
+    // aligned_alloc has undefined behavior if the length is not a multiple
+    // of the allignment.
```
Comment (matu3ba) [inline lib/std/heap.zig:186]: s/allignment/alignment

File: lib/std/heap.zig [hunk 2]
```diff
@@ -182,7 +182,10 @@ fn rawCAlloc(
...
+    const adjustedLen = if (len % @alignOf(std.c.max_align_t) == 0) len else (len + (@alignOf(std.c.max_align_t) - (len % @alignOf(std.c.max_align_t))));
```
Comment (matu3ba) [inline lib/std/heap.zig:187]: s/adjustedLen/adjusted_len

File: CMakeLists.txt [hunk 3]
```diff
@@ -158,6 +158,31 @@ if(APPLE AND ZIG_STATIC)
     list(APPEND LLVM_LIBRARIES "${CURSES}")
 endif()
 
+if(CMAKE_SYSTEM_NAME STREQUAL "FreeBSD" AND ZIG_STATIC)
```
Comment (matu3ba) [inline CMakeLists.txt:161]: This FreeBSD stuff belongs into another PR or the PR should explain the fix properly with problem, fix and outcome.

**Comment by matu3ba** [comment]
> There is a standard memory allocation function, aligned_alloc, that does exactly what is needed. The only caveat is that the length of the allocated memory needs to be a multiple of the alignment. Behavior is undefined if this is not true. That is the reason that I adjust the value of len if it isn't already a multiple of the alignment.

From https://en.cppreference.com/w/c/memory/aligned_alloc:
```
Passing a size which is not an integral multiple of alignment or an alignment which is not valid or not supported by the implementation causes the function to fail and return a null pointer (C11, as published, specified undefined behavior in this case, this was corrected by DR 460). Removal of size restrictions to make it possible to allocate small objects at restrictive alignment boundaries (similar to [alignas](https://en.cppreference.com/w/c/language/_Alignas)) has been proposed by [n2072](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2072.htm).
```
As I understand it, the unnecessary UB has been fixed in C17 in https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2244.htm#dr_460.

**Review by matu3ba — COMMENTED** [review]
The FreeBSD stuff, which feels like belongs into a separate PR with proper explanations.

> Malloc makes no guarantee about the alignment of memory that it returns.

From https://en.cppreference.com/w/c/memory/malloc:
"If allocation succeeds, returns a pointer that is suitably aligned for any object type with [fundamental alignment](https://en.cppreference.com/w/c/language/object#Alignment)."

>  The existing code does a run-time check to make sure that the proper alignment has been achieved and panics if it is not properly aligned.

How is aligned_alloc different other than returning NULL?

**Comment by LemonBoy** [comment]
The `raw_c_allocator` is/was meant to be a thin wrapper over `malloc` and `free`, to be used in cases where you're confident not to need weird alignments and hit that assertion. `c_allocator` is the real deal and, as specified in #6385, it uses `posix_memalign` where available or tries to emulate it using the plain old overallocating `malloc` trick (which is exactly what `aligned_alloc` does, see [musl's implementation](https://git.musl-libc.org/cgit/musl/tree/src/malloc/mallocng/aligned_alloc.c?h=v1.2.2) ).

**Comment by kimshrier** [comment]
The problem is that this assertion is being hit 100% of the time when running the stage3 compiler on FreeBSD.

A simple command showing the problem.
```
$ stage3/bin/zig build -h
thread 189880 panic: incorrect alignment
/home/kim/Projects/zig/from_github/zig/lib/std/heap.zig:185:65: 0x5a3d8e in rawCAlloc (zig)
    const ptr = @alignCast(@alignOf(std.c.max_align_t), c.malloc(len));
                                                                ^
/home/kim/Projects/zig/from_github/zig/lib/std/mem/Allocator.zig:86:34: 0x6053d7 in allocAdvancedWithRetAddr__anon_6134 (zig)
    return self.vtable.alloc(self.ptr, len, ptr_align, ret_addr);
                                 ^
/home/kim/Projects/zig/from_github/zig/lib/std/mem/Allocator.zig:192:41: 0x727bb5 in alignedAlloc__anon_40268 (zig)
    return self.allocAdvancedWithRetAddr(T, alignment, n, @returnAddress());
                                        ^
/home/kim/Projects/zig/from_github/zig/lib/std/array_list.zig:380:67: 0x72d9af in ensureTotalCapacityPrecise (zig)
                const new_memory = try self.allocator.alignedAlloc(T, alignment, new_capacity);
                                                                  ^
/home/kim/Projects/zig/from_github/zig/lib/std/array_list.zig:356:51: 0x6047dd in ensureTotalCapacity (zig)
            return self.ensureTotalCapacityPrecise(better_capacity);
                                                  ^
/home/kim/Projects/zig/from_github/zig/lib/std/array_list.zig:391:60: 0x5a4114 in ensureUnusedCapacity (zig)
            return self.ensureTotalCapacity(self.items.len + additional_count);
                                                           ^
/home/kim/Projects/zig/from_github/zig/lib/std/array_list.zig:229:48: 0x5a0309 in appendSlice (zig)
            try self.ensureUnusedCapacity(items.len);
                                               ^
/home/kim/Projects/zig/from_github/zig/lib/std/fs/path.zig:679:39: 0x61c463 in resolvePosix (zig)
                try result.appendSlice(component);
                                      ^
/home/kim/Projects/zig/from_github/zig/lib/std/fs/path.zig:458:28: 0x6289f6 in resolve (zig)
        return resolvePosix(allocator, paths);
                           ^
/home/kim/Projects/zig/from_github/zig/src/Module.zig:4978:35: 0x8b26fc in importFile (zig)
        cur_pkg_dir_path, cur_file.sub_file_path, "..", import_string,
                                  ^
/home/kim/Projects/zig/from_github/zig/src/Compilation.zig:3457:42: 0x9ac7e8 in workerAstGenFile (zig)
                break :blk mod.importFile(file, import_path) catch continue;
                                         ^
/home/kim/Projects/zig/from_github/zig/src/ThreadPool.zig:87:39: 0x7b9181 in runFn (zig)
            @call(.auto, func, closure.arguments);
                                      ^
/home/kim/Projects/zig/from_github/zig/src/ThreadPool.zig:126:28: 0x7468d0 in worker (zig)
            runFn(&run_node.data);
                           ^
/home/kim/Projects/zig/from_github/zig/lib/std/Thread.zig:393:13: 0x95211a in callFn__anon_86509 (zig)
            @call(.auto, f, args);
            ^
/home/kim/Projects/zig/from_github/zig/lib/std/Thread.zig:666:30: 0x74723c in entryFn (zig)
                return callFn(f, args_ptr.*);
                             ^
???:?:?: 0x80c623839 in ??? (???)
???:?:?: 0x0 in ??? (???)
Abort trap (core dumped)
```

It looks like the zig function alignedAlloc in std/mem/Allicator.zig is calling self.allocAdvancedWithRetAddr which calls self.rawAlloc which calls self.vtable.alloc.  It looks to me that the alloc function pointed to in self.vtable should not be rawCAlloc.  Is this where the fix should be made?  What would be the appropriate function to use here?

**Comment by Vexu** [comment]
> The problem is that this assertion is being hit 100% of the time when running the stage3 compiler on FreeBSD.

There is a check in `main.zig` that selects between `c_allocator` and `raw_c_allocator`, it sounds like FreeBSD should force `c_allocator`.
https://github.com/ziglang/zig/blob/7350f0d9b5aecaf98998bc6591a649d9a7b7091f/src/main.zig#L165-L170

**Comment by LemonBoy** [comment]
What's the `sizeof` and `alignof` (using GCC/Clang) of `max_align_t` on FreeBSD?

**Comment by The-King-of-Toasters** [comment]
It's `32` and `16` on x86_64 and aarch64.

**Comment by LemonBoy** [comment]
Eh, that's weird. If the alignment and size of `std.c.max_align_t` match the C counterpart I'd expect the `malloc`-d pointer to be 0x10 aligned and the `aligncast` to succeed.

File: src/main.zig [hunk 4]
```diff
@@ -162,6 +162,9 @@ pub fn main() anyerror!void {
         if (use_gpa) {
             break :gpa general_purpose_allocator.allocator();
         }
+        if (builtin.os.tag == .freebsd) {
+            break :gpa std.heap.c_allocator;
+        }
```
Comment (andrewrk) [inline src/main.zig:167]: please add a comment explaining why this is done

**Review by andrewrk — CHANGES_REQUESTED** [review]

**Pull request closed without merging**